src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java

```diff
@@ -151,91 +144,55 @@ public SkyValue compute(SkyKey skyKey, Environment env)
-      handler = skylarkHandler;
-    } else {
-      handler = handlers.get(rule.getRuleClass());
-    }
+    RepositoryFunction handler = getHandler(rule);
     if (handler == null) {
       // If we refer to a non repository rule then the repository does not exist.
       return RepositoryDirectoryValue.NO_SUCH_REPOSITORY_VALUE;
     }
 
-    handler.setClientEnvironment(clientEnvironmentSupplier.get());
-
     byte[] ruleSpecificData = handler.getRuleSpecificMarkerData(rule, env);
-    if (ruleSpecificData == null) {
-      return null;
-    }
-    String ruleKey = computeRuleKey(rule, ruleSpecificData);
-    Map<String, String> markerData = new TreeMap<>();
-     if (isFetch.get() && handler.isLocal(rule)) {
-      // Local repositories are fetched regardless of the marker file because the operation is
-      // generally fast and they do not depend on non-local data, so it does not make much sense to
-      // try to cache from across server instances.
-      setupRepositoryRoot(repoRoot);
-      env.getListener().post(new RepositoryFetching(repositoryName.getName(), false));
-      RepositoryDirectoryValue.Builder localRepo =
-          handler.fetch(rule, repoRoot, directories, env, markerData, skyKey);
-      if (localRepo == null) {
-        return null;
-      } else {
-        // We write the marker file for local repository essentially for getting the digest and
-        // injecting it in the RepositoryDirectoryValue.
-        byte[] digest = writeMarkerFile(markerPath, markerData, ruleKey);
-        env.getListener().post(new RepositoryFetching(repositoryName.getName(), true));
-        return localRepo.setDigest(digest).build();
-      }
-    }
-
-    // We check the repository root for existence here, but we can't depend on the FileValue,
-    // because it's possible that we eventually create that directory in which case the FileValue
-    // and the state of the file system would be inconsistent.
-
-    byte[] markerHash = isFilesystemUpToDate(markerPath, rule, ruleKey, handler, env);
     if (env.valuesMissing()) {
       return null;
     }
-    if (DONT_FETCH_UNCONDITIONALLY.equals(fetchUnconditionally)
-        && markerHash != null
-        && repoRoot.exists()) {
-      // Now that we know that it exists and that we should not fetch unconditionally, we can
-      // declare a Skyframe dependency on the repository root.
-      RepositoryFunction.getRepositoryDirectory(repoRoot, env);
-      if (env.valuesMissing()) {
-        return null;
+    DigestWriter digestWriter = new DigestWriter(directories, repositoryName, rule,
+        Preconditions.checkNotNull(ruleSpecificData));
+
+    // Local repositories are fetched regardless of the marker file because the operation is
+    // generally fast and they do not depend on non-local data, so it does not make much sense to
+    // try to cache from across server instances.
+    boolean fetchLocalRepositoryAlways = isFetch.get() && handler.isLocal(rule);
```

Comment:
suggestion: instead of calling `isFetch()` in both of the if statements (lint 164 and line 182) wrap them both in an `if (isFetch()) {` block and potentially extract its contents into a method called e.g. `maybeFetchRepository()`. I found the logic a little hard to follow here, especially due to all the negations (both with the `!` operator and negative variable names and maybe that would help.

Reply:
This is not possible, because in case of isFetch is false we are still checking if the repository already exists and we just return the existing value.

Reply:
Nevermind. I got confused about the De Morgan identities for a bit.

Reply:
My eyes are *still* bleeding about the number of negations here (`!fetch` and three negations in `!doNotFetchUnconditionally`). Not that I have much better idea.  Let's see. The condition below is:

`repoRoot.exists() && doNotFetchUnconditionally && !(isFetch.get() && isLocal)`

Applying a De Morgan identity we get:

`repoRoot.exists() && doNotFetchUnconditionally && (!isFetch.get() || !isLocal)`

so we do the up-to-dateness check if:

1. the root exists
2. unconditional fetching is not enabled
3. the repository is remote *or* fetching is disabled

Why do we skip the check is fetching is disabled? In that case, do we do anything at all?

Reply:
If the fetching is disabled, we still check if the repository is up-to-date and then we do nothing.

Reply:
Sorry, I think two-step logic is better here:
1) we check the category of the repository: local/other.
2) for other repositories, we check if they are already up-to-date.

I added an explaining comment.

Reply:
all the power to you!